### DISCUSSION TITLE
How to get encoding quality of jpeg image?

### DISCUSSION THREAD

**Discussion by stanley-zenfolio** [libvips/libvips#2089] [Q&A]
Ask a question...

**Comment by jcupitt** [comment]
Hi @stanley-zenfolio,

This is not possible, unfortunately, JPEG does not record encoding quality in the file.

**Comment by stanley-zenfolio** [comment]
Is there any document or paper describing how it can be calculated? I noticed that magick seems to use quantization to compute it internally and then exposing the value in property Quality.

**Comment by jcupitt** [comment]
You're right, imagemagick attempts to estimate it, though of course it won't always be correct. You'd need to read the IM source code to see the rule they use.

**Comment by lovell** [comment]
Please also see #895 and https://github.com/lovell/sharp/issues/908

**Comment by jcupitt** [comment]
Just for fun I tried plotting the compression ratio against the Q factor, and it's very close to a straight line;

![plot](https://user-images.githubusercontent.com/580843/102328411-f2fc9200-3f7e-11eb-9b39-2a19e70e95d0.png)

So I think you could estimate Q reasonably accurately from image dimensions + compressed size in a few lines of code.

**Comment by jcupitt** [comment]
Here's a tiny ruby program to generate a compression curve from an image:

```ruby
#!/usr/bin/ruby

require 'vips'

image = Vips::Image.new_from_file ARGV[0]
uncompressed_size = image.width * image.height * image.bands
(1..100).each do |q|
  data = image.jpegsave_buffer(Q: q, strip: true, optimize_coding: true)
  compression_ratio = uncompressed_size / data.length
  puts "#{q}, #{compression_ratio}"
end
```

For a photograph, I get:

![plot](https://user-images.githubusercontent.com/580843/102329958-e2e5b200-3f80-11eb-91ab-8bf385f751c4.png)

The trend line is a linear regression for Q>50, assuming no one is interested in very low values of Q (since the quality is horrible). The equation is almost the same.

**Comment by stanley-zenfolio** [comment]
Is the equation y = -0.573032x + 62.644 approximately apply to all jpegs or every jpeg has a different equation?

**Comment by Weissrolf** [comment]
Faststone Image Viewer displays JPEG quality and by default is set to save them at original quality settings. I always assumed that encoding quality is saved as part of the file because of this feature.

**Comment by jcupitt** [comment]
I think they probably guess the Q value from looking at the huffman tables (this is what imagemagick does).

Q is a hint to the encoder and doesn't have a precise meaning. Q80 on one encoder might be closest to Q70 on another.

**Comment by jtlz2** [comment]
Sherloq (python) estimates this from the tables - see e.g. https://github.com/GuidoBartoli/sherloq/blob/master/gui/quality.py
